### DISCUSSION TITLE
How do I pick a point label?

### DISCUSSION THREAD

**Discussion by danelinquist** [pyvista/pyvista#3976] [Q&A]
I am developing a GUI that uses PyVista for the visualization of an unstructured grid mesh. I am using point labels to display dimension values of the mesh and would like to be able to select the Dimension (point label) and update the Dimension Value (label text) through a callback routine.I have done similar things with the mesh, where I use the **enable_point_picking** method to select a node of the unstructured grid and return the node ID through a callback routine. 

From the documentation, a point label has the option to make the entity pickable, by setting the 'pickable' argument of **add_point_label** to **True**. What I can't find in the documentation, is what picking method (e.g., **enable_point_picking**, **enable_surface_picking**, **enable_mesh_picking**, etc.) is used to select the label. 

I have attached an image to better illustrate what I am trying to accomplish. Any help would be greatly appreciated.
![image](https://user-images.githubusercontent.com/71725627/217382350-70d206b9-ac59-428b-9f0c-481980089fc5.png)

**Comment by banesullivan** [comment]
This is tough.

You might be able to modify `enable_mesh_picking()` to do what you need since it leverages `vtkPropPicker`. You'll also need to make the labels "pickable"

Some psuedo code:

```py
import numpy as np
import pyvista as pv

mesh = pv.Cube()

def callback(actor):
    print(type(actor))

plotter = pv.Plotter(notebook=0)
plotter.add_mesh(mesh, pickable=False)

label_actor = plotter.add_point_labels([mesh.points[0]], ["My Label"], point_size=20, font_size=36, pickable=False)
plotter.remove_actor(label_actor)
plotter.add_actor(label_actor, reset_camera=False, name='label', pickable=True)

plotter.enable_mesh_picking(callback, use_actor=True)
plotter.show()
```

This isn't working, though and I don't know if `vtkActor2D` is easily pickable with the `vtkLabelPlacementMapper`

**Reply by banesullivan** [reply]
I'd recommend asking the VTK discourse if picking labels like this is possible in the first place https://discourse.vtk.org/

**Reply by danelinquist** [reply]
I dug into the source code and it looks like the '**pickable**' argument for **add_point_label** is only used if the '**show_points**' argument is also used. Adding a point label with both '**pickable**' and '**show_points**' set to **True** actually builds two actors:  a vtkActor2D for the label (name ='{name}-labels') and a PolyData point actor (name = '{name}-points').
If I enable these options when building the point labels for the different dimensions, I can select the point attached to the label (see Label T1-069 below). 

![image](https://user-images.githubusercontent.com/71725627/217686920-c2cd120f-2c0b-4390-8a77-be8c8f78ba2e.png)

Since both the point and the label share the same name prefix, I should be able to set up a callback routine that takes the name of the picked Point actor, removes the "-points" suffix, and changes it to "-labels" to point to the label instead. Once I have the label actor name, I should be able to modify the label text relatively easily. 

Reading through the documentation, I found the following example that modifies a picked Actor's properties in a similar way to what I am trying to do: https://docs.pyvista.org/examples/02-plot/mesh-picking.html .  It looks like the **enable_mesh_picking** method has a '**use_actors**' argument, that will return the PyVista actor through a callback, which allows the properties of the actor, such as color, to be modified, but when I try running this example, I get several errors.

Is this example still valid, or have I run into a bug?

**Reply by danelinquist** [reply]
Just wanted to provide an update since I see others with the same question.

As far as I can tell, labels themselves are not pickable through any current method in PyVista. When adding a point label and setting  both 'pickable' and 'show_points' to True, two actors are built: a vtkActor2D for the label (name ='{name}-labels') and a PolyData point actor (name = '{name}-points'). This of course, adds a visible point which is not desired in some cases, and the point properties such as size and color can't be directly changed.

To get around this limitation, I first added the point label:

self.plotter.add_point_labels(labelCenter, [labelText], **labelKwargs)
    where:
        labelKwargs = {
                        "font_size": 15, 
                        "show_points": True, 
                        "text_color": "black",
                        "always_visible": False,
                        "shape": None,
                        "fill_shape": False,
                        "margin": 3,
                        "name": labelName,
                        "pickable": True,
                        "tolerance": 0.001,
                    }

Next, we need to overwrite the point actor with modified properties. As previously noted, creating a point label with name "LabelName" actually creates two actors; the label actor with name "LabelName-labels" and the point actor with name "LabelName-points". To overwrite the point actor, create a new one with name "LabelName-points":

self.plotter.add_points(center, **pointKwargs)
    where:
        pointKwargs = {
                            "show_edges": False, # hide the edges
                            "color": "red", # color not important
                            "opacity": 0.001, # make the point almost transparent
                            "point_size": 50, # make the point big
                            "name": labelName + "-points",  # give it the same name as the previous point actor created
                            "pickable": True, # make it pickable
                        }
This will replace the auto generated point with a large transparent point at the same location as the label. Adjusting the "opacity" argument to make the points visible for clarity, we now have a pickable point that takes up around the same space as the label. 
![image](https://github.com/pyvista/pyvista/assets/71725627/a05f6e85-005d-4565-83c0-1391daab56eb)
Clicking on the label will now actually click on the point, which we can use to return the label actor by using the enable_mesh_picking method and a callback routine such as:
plotter.enable_mesh_picking(
                callback=labelClick,
                show=False,
                use_actor=True,
                left_clicking=True,
                show_message="Click on a label to modify.",
                color="black",
                font_size=10,
            )
    where:
        labelClick(actor):
        # Get the list of PyVista actors
        pointName = actor.name # this is the point actor name
        labelName = pointName .split("-points")[0] + '-labels' # this is the name of the label actor

Finally, setting the point opacity back to something small to make it almost transparent, we can see that the points aren't visible. In this example, the callback returns the label actor name which was used to change the text color from black to red.
![image](https://github.com/pyvista/pyvista/assets/71725627/b301fb22-16ec-4bbe-bcf9-6035f082a7a5)

This method works fairly well, but has some limitations as the point is anchored to the lower left corner of the label. In a case where there are many labels close together, it is easy to select the wrong label. With longer labels, the point may not cover the entire label, which can again cause selection issues.

**Reply by Indrajeetrt** [reply]
Hello @banesullivan @danelinquist with the above suggested method I am able to pick the node associated with the label, but my question is can we really move that label to somewhere other point? I tried to change the center of the label but couldn't do it as i encountered some error. can you suggest a solution for the concern?
